### ISSUE TITLE
在 finetune.py 代码里 labels 没有mask掉user的输入部分

### ISSUE THREAD

**Issue by ifromeast** [QwenLM/Qwen1.5#222]
```
    for i, msg in enumerate(messages):
        texts.append(
            tokenizer.apply_chat_template(
                msg,
                tokenize=True,
                add_generation_prompt=False,
                padding=True,
                max_length=max_len,
                truncation=True,
            )
        )
    input_ids = torch.tensor(texts, dtype=torch.int)
    target_ids = input_ids.clone()
    target_ids[target_ids == tokenizer.pad_token_id] = IGNORE_TOKEN_ID
    attention_mask = input_ids.ne(tokenizer.pad_token_id)
```
在微调数据集处理过程中，是按照pretrain 的处理方式，没有mask 掉labels中 system 和 user 的输入的部分，请问应该如何修改呢？

**Comment by jklj077** [comment]
Please also refer to https://github.com/QwenLM/Qwen1.5/issues/343

**Comment by disperaller** [comment]
> qwen1.5 采用了apply_chat_template()来构造input_ids，对encode之后的input想要修改label不是很方便，可以参考qwen的数据构造的preprocess()代码，我这里简单修改了下，供参考：
> 
> ```
> def preprocess(
>     sources,
>     tokenizer: transformers.PreTrainedTokenizer,
>     max_len: int,
>     system_message: str = None
> ) -> Dict:
>     """
>     2、system and user conversation token not take loss
>     """
>     roles = {"system":"<|im_start|>system", "user": "<|im_start|>user", "assistant": "<|im_start|>assistant"}
> 
>     im_start = tokenizer('<|im_start|>').input_ids
>     im_end = tokenizer('<|im_end|>').input_ids
>     nl_tokens = tokenizer('\n').input_ids
>     _system = tokenizer('system').input_ids + nl_tokens
>     _user = tokenizer('user').input_ids + nl_tokens
>     _assistant = tokenizer('assistant').input_ids + nl_tokens
> 
>     # Apply prompt templates
>     input_ids, targets = [], []
>     for i, source in enumerate(sources):
>         input_id, target = [], []
>         if system_message is not None:
>             if roles[source[0]["role"]] == roles["system"]:
>                 source = source[1:]
>             else:
>                 source = source
> 
>             system = im_start + _system + tokenizer(system_message).input_ids + im_end + nl_tokens
>             input_id += system
>             target += im_start + [IGNORE_TOKEN_ID] * (len(system)-3) + im_end + nl_tokens
>             assert len(input_id) == len(target)
> 
>         for j, sentence in enumerate(source):
>             role = roles[sentence["role"]]
>             _input_id = tokenizer(role).input_ids + nl_tokens + tokenizer(sentence["content"]).input_ids + im_end + nl_tokens
>             input_id += _input_id
> 
>             if role == roles["system"]:
>                 _target = im_start + [IGNORE_TOKEN_ID] * (len(_input_id)-3) + im_end + nl_tokens
>             elif role == roles["user"]:
>                 _target = im_start + [IGNORE_TOKEN_ID] * (len(_input_id)-3) + im_end + nl_tokens
>             elif role == roles["assistant"]:
>                 _target = im_start + [IGNORE_TOKEN_ID] * len(tokenizer(role).input_ids) + _input_id[len(tokenizer(role).input_ids)+1:-2] + im_end + nl_tokens
>             else:
>                 raise NotImplementedError
>             target += _target
>         assert len(input_id) == len(target)
>         input_id += [tokenizer.pad_token_id] * (max_len - len(input_id))
>         target += [IGNORE_TOKEN_ID] * (max_len - len(target))
>         input_ids.append(input_id[:max_len])
>         targets.append(target[:max_len])
> 
>     input_ids = torch.tensor(input_ids, dtype=torch.int)
>     targets = torch.tensor(targets, dtype=torch.int)
> 
>     return dict(
>         input_ids=input_ids,
>         target_ids=targets,
>         attention_mask=input_ids.ne(tokenizer.pad_token_id),
>     )
> ```

So the predefined tokens theoretically should be involved in the computation of loss? For example, the starting token `<|im_start|>`, should this token participate in the loss computation?

**Comment by STHSF** [comment]
I think it's up to you, the special token in Qwen and Qwen1.5 participated in training. but in chatglm3 not but some of it participate in training. different models have different strategy.

**Comment by disperaller** [comment]
> I think it's up to you, the special token in Qwen and Qwen1.5 participated in training. but in chatglm3 not but some of it participate in training. different models have different strategy.

how about llama3? should we include, for example, eos token in the loss computation when training the 8b base model?

**Comment by STHSF** [comment]
you can refer to the official SFT code
